ENH: Change DSSP `get_hbond_map` to use capped_distance and PBC-aware distance calculations

BradyAJohnston opened the pull request:
Changes made in this Pull Request:

This updates the `get_hbond_map` function to use `MDAnalysis.lib.distances.capped_distance`. This can now respect periodic boundary crossing for proper calculation of SS across boundaries. The `backend` option is now also supported for `distopia` and `openmp`. Nothing changes in the API except for the additional `backend` argument.

Rather than computing a pairwise distance matrix of every residue against every other residue, we can utilised KDTrees first with `capped_distance` to find potential pairs that also looks across boundaries and use `calc_bonds` for accelerated distance calculations that also respect boundaries.

The returned value from `get_hbond_map` remains the same but we have avoided having to compute distances and energies between residues that would never form an HBond to begin with.

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [ ] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5182.org.readthedocs.build/en/5182/

<!-- readthedocs-preview mdanalysis end -->

BradyAJohnston committed:
initial port to KDTree based HBond map

BradyAJohnston committed:
change distance_matrix -> query_ball_tree

BradyAJohnston committed:
tweak comments

BradyAJohnston committed:
tweak commments and layout

BradyAJohnston committed:
run black

BradyAJohnston committed:
add name to AUTHORS

BradyAJohnston committed:
add changelog and return comment

BradyAJohnston committed:
add test for donor_mask

BradyAJohnston committed:
run black

IAlibay commented on package/CHANGELOG:
```diff
@@ -32,6 +32,8 @@ Fixes
 Enhancements
  * Enables parallelization for analysis.diffusionmap.DistanceMatrix
    (Issue #4679, PR #4745)
+ * Change `get_hbond_map` to use KDTree for finding nearby potential H bonding pairs
```
New entries at the top please :)

IAlibay commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -176,52 +178,54 @@ def get_hbond_map(
-    hmap = np.tile(h_1, (1, 1, n)).reshape(n, n, 3)
+    # We can use KDTree to find candidate pairs within cutoff distance without having to
+    # compute a full pairwise distance matrix, much faster especially for larger proteins
+    n_kdtree = KDTree(n_atoms)
```
I know the original code was vendored from somewhere else, so it explains the heavy reliance on `linalg.norm` and the lack of PBC awareness - however, now that we're making changes, *should* we be using one of our `lib.distances` calls (maybe `capped_distance`) instead?

BradyAJohnston replied:
I'm not at all experienced with `lib.distances` are you talking about for the KDTree / potential pair generation or for the actual distance computation? If just for computing distances then on the KDTree generation we would already be discarding some potential across-boundary bonds.

We should be able to provide _simple_ periodic checking with [`KDTree`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html) you can specify a `boxsize` parameter. But this wouldn't take into account the more complex boundaries we might come across?

BradyAJohnston replied:
Okay I just dug in and found `capped_distance` which should do exactly what we are after with better periodic boundary checking.

BradyAJohnston replied:
The performance isn't _as good_ but also as expected with more checks. Still a large improvement over current.

```
============================================================
Scaling benchmark:
Residues     Time (ms)       Time/Res (μs)
------------------------------------------------------------
50           0.142           2.835
100          0.291           2.906
150          0.489           3.257
200          0.736           3.682
250          1.056           4.226
300          1.629           5.429
============================================================
```

BradyAJohnston replied:
I've currently updated the pair search to use `capped_distances` - but the actual distance computation still isn't taking into account periodic boundaries. Is there a periodic boundary aware distance calculation that doesn't do "everything vs everything" which `distance_array` seems to do? Ideally we would avoid repeatedly calling `capped_distance` as we have already constructed the KDTree once and can just use those results for computation later on, but still taking into account periodic boundary crossing.

orbeckst replied:
Pair-wise distances between two equal groups of atoms can be calculated with [MDAnalysis.lib.distances.calc_bonds](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.calc_bonds).

BradyAJohnston committed:
change KDTree -> capped_distances

BradyAJohnston committed:
run black

BradyAJohnston committed:
move changelog entry

marinegor commented:
@BradyAJohnston wow, that's super impressive results, thanks! Some comments:

1. 

>but the actual distance computation still isn't taking into account periodic boundaries. Is there a periodic boundary aware distance calculation that doesn't do "everything vs everything" which distance_array seems to do?

Just to make sure I understand this correctly: the current implementation is plain wrong when the boundary condition is crossed, since it just computes the wrong distance, right? I mean this part:

```python
    # compute distances and energy as previously but only for the potential pairs
    # still returning the same energy matrix that would have otherwise been made
    o_indices = pairs[:, 1]
    n_indices = pairs[:, 0]
    d_on = np.linalg.norm(o_atoms[o_indices] - n_atoms[n_indices], axis=-1)
    d_ch = np.linalg.norm(c_atoms[o_indices] - h_1[n_indices], axis=-1)
    d_oh = np.linalg.norm(o_atoms[o_indices] - h_1[n_indices], axis=-1)
    d_cn = np.linalg.norm(c_atoms[o_indices] - n_atoms[n_indices], axis=-1)
```

so we're looking for periodic contition aware `np.linalg.norm` replacement? Or can we just put `capped_distance` or smth instead of `-`, like

```python
    d_on = np.linalg.norm(capped_distance(o_atoms[o_indices], n_atoms[n_indices]))
    ...
```


2.

When defaulting to `self._trajectory.ts.dimensions`, I'd perhaps consider changing / removing it if the structure comes from RCSB -- cryoEM structures default to some hardcoded unit cell which can even be too small, so it might break things.

BradyAJohnston commented:
Hi @marinegor yes if `DSSP` is currently run on a protein that has secondary structure over the periodic cell then it would incorrectly _not_ assign it. 

We could just have multiple `capped_distance` calls which would solve the issue with periodic crossing, but each time we would be re-building a KDTree when ideally we can just re-use the pairs from the first `capped_distance` call which gives the relevant pairs to check already.

If there was just a `periodic_distance(c_atoms[o_indices] - h_1[n_indices])` that we can call then we compute the distance and allow for periodic crossing. 

In the current setup while we do _detect_ periodic bonds to check - the results end up being wrong like you say.

marinegor commented:
>If there was just a periodic_distance(c_atoms[o_indices] - h_1[n_indices]) that we can call then we compute the distance and allow for periodic crossing.

perhaps there are some functions in distopia (namely, `distopia.distances_ortho` that you could use directly, or am I wrong: https://www.mdanalysis.org/distopia/api)


>In the current setup while we do detect periodic bonds to check - the results end up being wrong like you say.

do you see this being a separate issue, perhaps if you don't find a quick solution to use among `distopia` library? I feel like this PR doesn't have to be blocked by this discussion.

IAlibay commented:
> If there was just a periodic_distance(c_atoms[o_indices] - h_1[n_indices]) that we can call then we compute the distance and allow for periodic crossing.

@BradyAJohnston have a look at `minimize_vectors` https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/distances.py#L2106

IAlibay commented:
> > If there was just a periodic_distance(c_atoms[o_indices] - h_1[n_indices]) that we can call then we compute the distance and allow for periodic crossing.
> 
> perhaps there are some functions in distopia (namely, `distopia.distances_ortho` that you could use directly, or am I wrong: https://www.mdanalysis.org/distopia/api)

Please don't try to use distopia directly, it's meant to be an (optional for now) backend replacement.

BradyAJohnston commented:
Thanks for the suggestion @IAlibay - this would then just become as such:

```py
    d_on = np.linalg.norm(minimize_vectors(o_atoms[o_indices] - n_atoms[n_indices], box=box), axis=-1)
    d_ch = np.linalg.norm(minimize_vectors(c_atoms[o_indices] - h_1[n_indices], box=box), axis=-1)
    d_oh = np.linalg.norm(minimize_vectors(o_atoms[o_indices] - h_1[n_indices], box=box), axis=-1)
    d_cn = np.linalg.norm(minimize_vectors(c_atoms[o_indices] - n_atoms[n_indices], box=box), axis=-1)
```

The usage of `np.linalg.norm` should now be fine we are taking into account the periodic crossing?

marinegor commented:
>The usage of np.linalg.norm should now be fine we are taking into account the periodic crossing?

I guess so, yeah -- it's just a norm calculation, there should not be any problem with that as long as the vector itself is boundary-corrected.

As for cryoEM default, the placeholder unit cell is `(1, 1, 1, 90, 90, 90)`, so you can explicitly compare with that and e.g. default to no box calculations if it's set.

UPD: added this remark to changes request.

marinegor commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -399,7 +399,11 @@ def _get_coords(self) -> np.ndarray:
+        dssp = assign(
+            coords,
+            donor_mask=self._donor_mask,
+            box=self._trajectory.ts.dimensions,
```
if dimensions are `(1, 1, 1, 90, 90, 90)` (cryoEM placeholder), just set to `None`.

IAlibay replied:
This shouldn't be done here, but rather when loading the PDB - there is an a bad edge case where you might actually just have that unit box.

IAlibay replied:
Indeed we do this already: https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/PDB.py#L228-L231

IAlibay replied:
On further thought, I wouldn't be completely against this check - maybe a warning would be ok too if we don't want to be too strict

IAlibay replied:
On further thought, I wouldn't be completely against this check - maybe a warning would be ok too if we don't want to be too strict

BradyAJohnston replied:
Not 100% with how this should be handled. If I am raising a warning in _single_frame won't that warning be raised a bunch of times for every frame?

BradyAJohnston replied:
```py
def _single_frame(self):
        coords = self._get_coords()
        DEFAULT_BOX_DIMENSIONS = (1, 1, 1, 90, 90, 90)
        box = self._trajectory.ts.dimensions

        if box == DEFAULT_BOX_DIMENSIONS:
            box = None
            warnings.warn(
                f"Default box dimensions {DEFAULT_BOX_DIMENSIONS} are treated as no box"
            )
        dssp = assign(
            coords,
            donor_mask=self._donor_mask,
            box=self._trajectory.ts.dimensions,
        )
        self.results.dssp_ndarray.append(dssp)
```

IAlibay replied:
@BradyAJohnston - how about making that warning happen in init? The atomgroups are set there, so it would be best to do it when the class is created.

IAlibay commented:
> 
> ```python
>     d_on = np.linalg.norm(minimize_vectors(o_atoms[o_indices] - n_atoms[n_indices], box=box), axis=-1)
>     d_ch = np.linalg.norm(minimize_vectors(c_atoms[o_indices] - h_1[n_indices], box=box), axis=-1)
>     d_oh = np.linalg.norm(minimize_vectors(o_atoms[o_indices] - h_1[n_indices], box=box), axis=-1)
>     d_cn = np.linalg.norm(minimize_vectors(c_atoms[o_indices] - n_atoms[n_indices], box=box), axis=-1)
> ```
> 
> The usage of `np.linalg.norm` should now be fine we are taking into account the periodic crossing?

Yep, don't trust me fully (I'm sleep deprived), but that sounds reasonable.

marinegor commented:
One thing, it probably shouldn't happen in `_single_frame` but rather in `__init__`, otherwise i'm ok with a warning.

BradyAJohnston committed:
use  before distance calculations

BradyAJohnston committed:
warning for init of default box dimensions

BradyAJohnston commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -323,6 +324,12 @@ def __init__(
+        if np.array_equal(self._box, (1, 1, 1, 90, 90, 90)):
+            self._box = None
+            warnings.warn(
+                "Box dimensions are (1, 1, 1, 90, 90, 90), not using periodic boundary conditions in DSSP calculations"
```
Not 100% on what the warning should be - but this is how we are imagining it? @IAlibay

IAlibay replied:
Yes that looks about right. With regards to the warning, your message is close enough to what I was thinking, i.e. something like "Default unit cells of 1 A cubed found. This likely comes from a cryo-em file. To avoid usses, we are disabling periodic boundary conditions in the DSSP calculation".

BradyAJohnston replied:
A warning is currently already raised twice: once when loading the universe and again when running the DSSP (not sure why it is happening for the `run()`. As the unit cell dimensions are already set to `None` it's not triggering the new warning. In most cases the unit cell wouldn't make it "all the way" to DSSP but I'll change it to just be the same warning message.

```py
In [9]: p = Path("/Users/brady/Downloads/9N9M.pdb")

In [10]: u = mda.Universe(p)
/Users/brady/git/mdanalysis/package/MDAnalysis/coordinates/PDB.py:479: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn(

In [11]: a = dssp.DSSP(u)

In [12]: a.run()
/Users/brady/git/mdanalysis/package/MDAnalysis/coordinates/PDB.py:479: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn(
/Users/brady/git/mdanalysis/package/MDAnalysis/analysis/base.py:562: UserWarning: Reader has no dt information, set to 1.0 ps
  self.times[idx] = ts.time

Out[12]: <MDAnalysis.analysis.dssp.dssp.DSSP at 0x110d4c910>
```

IAlibay replied:
Yeah this is what I was trying to convey earlier in the conversation when I was saying that the PDB reader takes care of that edge case already. I'm not sure what the best approach is here, thoughts @marinegor ?

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -176,58 +190,70 @@ def get_hbond_map(
+            return np.linalg.norm(x[o_indices] - y[n_indices], axis=-1)
+        else:
+            return np.linalg.norm(
+                minimize_vectors(x[o_indices] - y[n_indices], box=box),
```
or use directly [MDAnalysis.lib.distances.calc_bonds](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.calc_bonds)
```python
return calc_bonds(x[o_indices], y[n_indices], box=box)
```

BradyAJohnston replied:
Ah this is good to know! (and yeah I missed it from naming). Will update to use that instead of this combo.

orbeckst replied:
use `calc_bonds` instead of `norm(minimize_vectors)`

orbeckst replied:
add `backend=backend`

orbeckst reviewed (commented):
Good idea to use our distance functions. (see comments on using [calc_bonds](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.calc_bonds), whose name betrays its flexibility and thus may not be widely know as the _the_ function for PBC-aware distances.

I would add an explicit `backend` keyword argument to all calls of our distance functions (and a way for the user to also set the backend when invoking the class) so that it becomes possible to enable parallel/distopia backends.

orbeckst commented:
@marinegor may I assign you as the PR shepherd? If you don't have the bandwidth, please unassign yourself.

orbeckst commented on package/CHANGELOG:
```diff
@@ -30,6 +30,8 @@ Fixes
    DSSP by porting upstream PyDSSP 0.9.1 fix (Issue #4913)
 
 Enhancements
+ * Change `get_hbond_map` to use `distance_capped` to speedup finding H bonding pairs
```
This should make clear that we're fixing an issue in DSSP (and also speeding it up)

orbeckst commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -323,6 +324,12 @@ def __init__(
         }
         self._donor_mask: Optional[np.ndarray] = ag.residues.resnames != "PRO"
+        self._box = self._trajectory.ts.dimensions
+        if np.array_equal(self._box, (1, 1, 1, 90, 90, 90)):
```
Should be [`np.allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html) to do proper floating point comparison.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -11,10 +11,13 @@
 
 import numpy as np
 
+from MDAnalysis.lib.distances import capped_distance, minimize_vectors
```
use `calc_bonds` instead of `norm(minimize_vectors)`

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -145,6 +149,14 @@ def get_hbond_map(
+        ``[lx, ly, lz, alpha, beta, gamma]``.
+
+         .. versionadded:: 2.11.0
+
```
check that it's rendered properly, might need an additional (second) blank line

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -156,6 +168,8 @@ def get_hbond_map(
        Support masking of hydrogen donors via `donor_mask` (especially needed
        for ignoring HN on proline residues). Backport of PRO fix from pydssp 0.9.1.
+    .. versionchanged:: 2.11.0
+       Speedup with KDTree and support periodic boundary checking via `box` param.
```
update, not KDTree

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -176,58 +190,70 @@ def get_hbond_map(
+                axis=-1,
+            )
+
+    d_on = _distances(o_atoms, n_atoms, box)
```
`d_on` has already been calculated by capped_distances, I don't think you need to recalculate.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -118,10 +121,11 @@ def _get_hydrogen_atom_position(coord: np.ndarray) -> np.ndarray:
     cutoff: float = DEFAULT_CUTOFF,
     margin: float = DEFAULT_MARGIN,
     return_e: bool = False,
+    box: np.ndarray | None = None,
```
pass `backend` through to the distance calculations so that it becomes possible to select `backend="distopia"` at the `DSSP` level

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -176,58 +190,70 @@ def get_hbond_map(
+        n_atoms,
+        o_atoms,
+        max_cutoff=HBOND_SEARCH_CUTOFF,
+        box=box,
```
add `backend=backend`

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -257,10 +291,12 @@ def assign(
     .. versionchanged:: 2.10.0
        Support masking of donors.
+    .. versionchanged:: 2.11.0
+       Speedup with KDTree and support periodic boundary checking via `box` param.
```
update

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -240,12 +266,20 @@ def assign(
+        ``[lx, ly, lz, alpha, beta, gamma]``.
+
+        .. versionadded:: 2.11.0
+
```
pass through `backend`

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_dssp.py:
```diff
@@ -35,6 +36,19 @@ def test_trajectory(client_DSSP):
+
+    assert not np.array_equal(
+        assign(coords), assign(coords, donor_mask=dssp._donor_mask)
+    )
```
Can we also assert something concrete (eg array shape)?

orbeckst reviewed (changes requested):
Use `calc_bonds` and pass through `backend=...` from `DSSP` down to `get_hbond_map` so that users could opt to choose eg distopia.

BradyAJohnston committed:
Merge branch 'develop' into dssp-kdtree

BradyAJohnston committed:
address PR feedback

- More tests for donor_mask
- change KDtree -> capped_distance
- expose option for distopia backend

BradyAJohnston committed:
black format testsuite

BradyAJohnston committed:
fix is_proline mask (bool instead of idx)

BradyAJohnston committed:
test coverage for box dimensions warning

BradyAJohnston commented:
I tried setting up some asv benchmarks but it seems like the timing actually isn't changing much at all (with the benchmarks I was making). I might try playing around with it some more for the PR description but otherwise this is ready to go I think.

orbeckst commented on package/CHANGELOG:
```diff
@@ -41,15 +41,21 @@ Fixes
    DSSP by porting upstream PyDSSP 0.9.1 fix (Issue #4913)
+ * DSSP now correctly handles periodic boundary conditions in hydrogen bond
+   detection; previously, hydrogen bonds across periodic boundaries were
+   missed (PR #5182)
```
Customarily, we add new items at the top.

(I am just saying it because I said so in every other PR I reviewed over the last few weeks...)

orbeckst commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -236,6 +237,12 @@ class DSSP(AnalysisBase):
 
+    backend : str, optional
+        Backend for distance calculations, by default ``"serial"``.
+        Can be set to ``"distopia"`` if `distopia`_ is installed.
```
link to distopia does not resolve

orbeckst commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -236,6 +237,12 @@ class DSSP(AnalysisBase):
+        Can be set to ``"distopia"`` if `distopia`_ is installed.
+
+        .. versionadded:: 2.11.0
+
```
```suggestion

    .. _distopia: https://www.mdanalysis.org/distopia/

```

orbeckst commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -236,6 +237,12 @@ class DSSP(AnalysisBase):
            for your specific case.
 
+    backend : str, optional
+        Backend for distance calculations, by default ``"serial"``.
```
Link to https://mdanalysis--5182.org.readthedocs.build/en/5182/documentation_pages/lib/distances.html#selection-of-acceleration-backend 
```suggestion
        :ref:`Backend for distance calculations<selection-of-acceleration-backend>`, by default ``"serial"``.
```

BradyAJohnston replied:
I haven't written these kind of docs before - do you have a good link for it? I've basically just always done markdown.

orbeckst replied:
* in the user guide: https://userguide.mdanalysis.org/stable/contributing_code.html#guidelines-for-writing-docstrings
* our user guide should really say explicitly that we are using restructured text with sphinx; https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html is a good intro
* for linking: cross referencing https://www.sphinx-doc.org/en/master/usage/referencing.html
* for specific roles for python code, look at the py-domain https://www.sphinx-doc.org/en/master/usage/domains/python.html (which is the default, this is from whence constructs like the class role in

  ```reST
  The :class:`~MDAnalysis.core.universe.Universe` ties a topology and a 
  trajectory together.
  ```

  come from.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -145,6 +150,20 @@ def get_hbond_map(
+
+        .. versionadded:: 2.11.0
+
+    backend : str, optional
+        Backend for distance calculations, by default ``"serial"``.
+        Can be set to ``"distopia"`` if `distopia`_ is installed.
+
+        .. versionadded:: 2.11.0
```
See comments above for linking/rendering.

This one is not even currently included in the docs (perhaps it should be?) so you could just remove the reST link from `distopia_`.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -240,12 +273,26 @@ def assign(
+
+        .. versionadded:: 2.11.0
+
+    backend : str, optional
+        Backend for distance calculations, by default ``"serial"``.
+        Can be set to ``"distopia"`` if `distopia`_ is installed.
+
+        .. versionadded:: 2.11.0
```
See comments above on linking. Probably just remove reST linking from `distopia_` or mark it a package but don't worry about linking:
```suggestion
    backend : str, optional
        Backend for distance calculations, by default ``"serial"``.
        Can be set to ``"distopia"`` if :mod:`distopia` is installed.

        .. versionadded:: 2.11.0
```

orbeckst reviewed (changes requested):
Minor docs issue (see comments).

If you have an ASV benchmark written, feel free to include it.

BradyAJohnston committed:
Update package/MDAnalysis/analysis/dssp/dssp.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

BradyAJohnston committed:
Update package/MDAnalysis/analysis/dssp/dssp.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

BradyAJohnston committed:
remaining docs comments

BradyAJohnston committed:
asv benchmark for dssp

BradyAJohnston committed:
Merge branch 'develop' into dssp-kdtree

orbeckst commented on benchmarks/benchmarks/analysis/dssp.py:
Thanks for including the benchmark.

Idea for the future: test different backends via test parameters; I tried looking into it but was getting stymied by lack of distopia conda-forge packages for Python 3.14 https://github.com/MDAnalysis/distopia/issues/190 — until we have those it won't run cleanly in our nightly benchmarking environment that looks at 3.11 and 3.14 https://github.com/MDAnalysis/benchmarks/blob/32322d3cc11167f3d5cb5a493db2bfcc2703de43/config/asv_c3potato.conf.json#L36

BradyAJohnston replied:
Yeah I initially tried against different backends but ran into issues with distopia & I haven't had any experience with it so left it. Glad it wasn't just me!

orbeckst replied:
We now have full distopia packages up to Python 3.14 (including macOS arm64). Might be worthwhile retrying eventuallu.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_dssp.py:
```diff
@@ -35,6 +38,65 @@ def test_trajectory(client_DSSP):
+# ensure that we get different assigned results when using and not using the
+# donor mask which filters out prolines from being potential HBond donors as they
+# are missing hydrogens
+def test_donor_mask(client_DSSP):
```
I don't think we need `client_DSSP` here.
```suggestion
def test_donor_mask():
```

orbeckst reviewed (approved):
Thanks for the changes and the benchmark file. 

(There's one minor fix that I'll quickly merge in.)

orbeckst committed:
Update testsuite/MDAnalysisTests/analysis/test_dssp.py

orbeckst commented:
@marinegor do you have the bandwidth to look at this PR? In particular, I think your comment was addressed by having the box warning in init.

orbeckst commented:
@IAlibay can you have a quick look at the PR and see if you're happy with the changes?

orbeckst committed:
Merge branch 'develop' into dssp-kdtree

orbeckst commented:
@IAlibay and @marinegor do you need more time to look at the PR or are you happy with me to merge?

orbeckst committed:
Merge branch 'develop' into dssp-kdtree

orbeckst committed:
Merge branch 'develop' into dssp-kdtree

orbeckst commented:
@marinegor and @IAlibay I think that the PR is functional as is and the warning issue that you both commented on was addressed. I haven't heard from both of you so I am assuming you're busy and will dismiss your reviews as stale. Thank you for reviewing the PR!

orbeckst committed:
Update CHANGELOG

IAlibay commented:
Ah apologies, this completely slipped by me. This looks great!
